modules/siddhi-core/src/main/java/io/siddhi/core/table/record/AbstractQueryableRecordTable.java

```diff
@@ -18,59 +18,585 @@
+                                                            MatchingMetaInfoHolder storeMatchingMetaInfoHolder,
+                                                            SiddhiQueryContext siddhiQueryContext,
+                                                            List<VariableExpressionExecutor>
+                                                                    storeVariableExpressionExecutors) {
+        MetaStateEvent metaStateEvent = new MetaStateEvent(storeMatchingMetaInfoHolder.getMetaStateEvent().
+                getMetaStreamEvents().length);
+        for (MetaStreamEvent referenceMetaStreamEvent: storeMatchingMetaInfoHolder.getMetaStateEvent().
+                getMetaStreamEvents()) {
+            metaStateEvent.addEvent(referenceMetaStreamEvent);
+        }
+        MatchingMetaInfoHolder matchingMetaInfoHolder = new MatchingMetaInfoHolder(
+                metaStateEvent,
+                storeMatchingMetaInfoHolder.getMatchingStreamEventIndex(),
+                storeMatchingMetaInfoHolder.getStoreEventIndex(),
+                storeMatchingMetaInfoHolder.getMatchingStreamDefinition(),
+                cacheTableDefinition,
+                storeMatchingMetaInfoHolder.getCurrentState());
+
+        Map<String, Table> tableMap = new ConcurrentHashMap<>();
+        tableMap.put(cacheTableDefinition.getId(), cachedTable);
+
+        return cachedTable.compileCondition(condition, matchingMetaInfoHolder,
+                storeVariableExpressionExecutors, tableMap, siddhiQueryContext);
+    }
+
+    protected abstract void connect() throws ConnectionUnavailableException;
 
     @Override
     public StreamEvent query(StateEvent matchingEvent, CompiledCondition compiledCondition,
                              CompiledSelection compiledSelection) throws ConnectionUnavailableException {
         return query(matchingEvent, compiledCondition, compiledSelection, null);
     }
 
+    public StreamEvent findInCache(CompiledCondition compiledCondition, StateEvent matchingEvent) {
+        if (isCacheEnabled) {
+            return cachedTable.find(matchingEvent, compiledCondition);
+        } else {
+            return null;
+        }
+    }
+
+    private Object[] appendValue(Object[] obj, Object newObj) {
+
+        ArrayList<Object> temp = new ArrayList<Object>(Arrays.asList(obj));
+        temp.add(newObj);
+        return temp.toArray();
+
+    }
+
+    @Override
+    public void add(ComplexEventChunk<StreamEvent> addingEventChunk) throws ConnectionUnavailableException {
+        if (isCacheEnabled) {
+            ComplexEventChunk<StreamEvent> addingEventChunkWithTimestamp = new ComplexEventChunk<>();
+            while (addingEventChunk.hasNext()) {
+                StreamEvent event = addingEventChunk.next();
+                Object[] outputData = event.getOutputData();
+                Object[] outputDataWithTimeStamp = appendValue(outputData,
+                        new Timestamp(System.currentTimeMillis()).getTime());
```

Comment:
Use siddhi's timestamp, you can get that from the context. do not use System.currentTimeMillis()

Reply:
ack